<a href="https://colab.research.google.com/github/Neuralnomad-hub/Machine-Learninng-Algorithms-implemented-in-datasets/blob/main/code_decisiontree.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

we are going to implement decision tree a;lgorithm into a classification dataset Titanic

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier,plot_tree
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import matplotlib.pyplot as plt

In [1]:
import seaborn as sns
df= sns.load_dataset('titanic')
print(df.head())


   survived  pclass     sex   age  ...  deck  embark_town  alive  alone
0         0       3    male  22.0  ...   NaN  Southampton     no  False
1         1       1  female  38.0  ...     C    Cherbourg    yes  False
2         1       3  female  26.0  ...   NaN  Southampton    yes   True
3         1       1  female  35.0  ...     C  Southampton    yes  False
4         0       3    male  35.0  ...   NaN  Southampton     no   True

[5 rows x 15 columns]


In [3]:
import pandas as pd
pd.set_option('display.max_columns', None)
print(df.head())

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  


In [4]:
#lets preprcess the the dataset first to the sex column by label encodig
from sklearn.preprocessing import LabelEncoder
le= LabelEncoder()
df['sex']= le.fit_transform(df['sex'])
print(df.head())

   survived  pclass  sex   age  sibsp  parch     fare embarked  class    who  \
0         0       3    1  22.0      1      0   7.2500        S  Third    man   
1         1       1    0  38.0      1      0  71.2833        C  First  woman   
2         1       3    0  26.0      0      0   7.9250        S  Third  woman   
3         1       1    0  35.0      1      0  53.1000        S  First  woman   
4         0       3    1  35.0      0      0   8.0500        S  Third    man   

   adult_male deck  embark_town alive  alone  
0        True  NaN  Southampton    no  False  
1       False    C    Cherbourg   yes  False  
2       False  NaN  Southampton   yes   True  
3       False    C  Southampton   yes  False  
4        True  NaN  Southampton    no   True  


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    int64   
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(5), object(4)
memory usage: 80.7+ KB


In [9]:
df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

Based on the output from df.isnull().sum(), we know:

'age' has 177 missing values.
'embarked' has 2 missing values.
'deck' has 688 missing values (a significant amount).
For the 'age' column, a common strategy is to fill missing values with the mean or median of the column. The mean is sensitive to outliers, while the median is more robust. Let's use the mean for now.

For the 'embarked' column, since it's a categorical column and only has a few missing values, we can fill them with the mode (the most frequent value).

For the 'deck' column, with such a high percentage of missing values, it might be best to drop this column entirely as it might not provide useful information and could even introduce noise.

In [13]:
df['age'].fillna(df['age'].median(), inplace=True)
df['embarked'].fillna(df['embarked'].mode()[0], inplace=True)
print(df.isnull().sum())

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64


/tmp/ipython-input-3840250118.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['age'].fillna(df['age'].median(), inplace=True)
/tmp/ipython-input-3840250118.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try 

In [14]:
df['embark_town'].fillna(df['embark_town'].mode()[0],inplace=True)

/tmp/ipython-input-751922478.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['embark_town'].fillna(df['embark_town'].mode()[0],inplace=True)


In [15]:
print(df.isnull().sum())

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64


In [17]:
#now only thing remained is preprocessing categorical columns
df['who']= le.fit_transform(df['who'])
df['adult_male']= le.fit_transform(df['adult_male'])
df['alive']= le.fit_transform(df['alive'])
df['alone']= le.fit_transform(df['alone'])
print(df.head())

   survived  pclass  sex   age  sibsp  parch     fare embarked  class  who  \
0         0       3    1  22.0      1      0   7.2500        S  Third    1   
1         1       1    0  38.0      1      0  71.2833        C  First    2   
2         1       3    0  26.0      0      0   7.9250        S  Third    2   
3         1       1    0  35.0      1      0  53.1000        S  First    2   
4         0       3    1  35.0      0      0   8.0500        S  Third    1   

   adult_male  embark_town  alive  alone  
0           1  Southampton      0      0  
1           0    Cherbourg      1      0  
2           0  Southampton      1      1  
3           0  Southampton      1      0  
4           1  Southampton      0      1  


In [20]:
df.head()


survived  pclass  sex   age  sibsp  parch     fare embarked  class  who  \
0         0       3    1  22.0      1      0   7.2500        S  Third    1   
1         1       1    0  38.0      1      0  71.2833        C  First    2   
2         1       3    0  26.0      0      0   7.9250        S  Third    2   
3         1       1    0  35.0      1      0  53.1000        S  First    2   
4         0       3    1  35.0      0      0   8.0500        S  Third    1   

   adult_male  embark_town  alive  alone  
0           1  Southampton      0      0  
1           0    Cherbourg      1      0  
2           0  Southampton      1      1  
3           0  Southampton      1      0  
4           1  Southampton      0      1

In [21]:
df.tail()

survived  pclass  sex   age  sibsp  parch   fare embarked   class  who  \
886         0       2    1  27.0      0      0  13.00        S  Second    1   
887         1       1    0  19.0      0      0  30.00        S   First    2   
888         0       3    0  28.0      1      2  23.45        S   Third    2   
889         1       1    1  26.0      0      0  30.00        C   First    1   
890         0       3    1  32.0      0      0   7.75        Q   Third    1   

     adult_male  embark_town  alive  alone  
886           1  Southampton      0      1  
887           0  Southampton      1      1  
888           0  Southampton      0      0  
889           1    Cherbourg      1      1  
890           1   Queenstown      0      1

In [23]:
#based on this categorical data we gonna do one hot encoding on embark_town,class,embarked
df= pd.get_dummies(df,columns=['embark_town','class','embarked'])
print(df.head())

   survived  pclass  sex   age  sibsp  parch     fare  who  adult_male  alive  \
0         0       3    1  22.0      1      0   7.2500    1           1      0   
1         1       1    0  38.0      1      0  71.2833    2           0      1   
2         1       3    0  26.0      0      0   7.9250    2           0      1   
3         1       1    0  35.0      1      0  53.1000    2           0      1   
4         0       3    1  35.0      0      0   8.0500    1           1      0   

   alone  embark_town_Cherbourg  embark_town_Queenstown  \
0      0                  False                   False   
1      0                   True                   False   
2      1                  False                   False   
3      0                  False                   False   
4      1                  False                   False   

   embark_town_Southampton  class_First  class_Second  class_Third  \
0                     True        False         False         True   
1                 